pkg/products/rhssouser/reconciler_multitenant.go

```diff
@@ -0,0 +1,710 @@
+	for _, realm := range realms {
+		if realm.Name == "master" || realm.Name == StagingRealmName {
+			continue
+		}
+		if !userExistsForRealm(mtUsers, realm) {
+			deleted = append(deleted, realm)
+		}
+	}
+
+	return added, deleted, nil
+}
+
+func userInList(added []userHelper.MultiTenantUser, user userHelper.MultiTenantUser) bool {
+	for _, usr := range added {
+		if usr.UID == user.UID {
+			return true
+		}
+	}
+	return false
+}
+
+// The loginEventUserId is the keycloak user id.
+// We can use this to find the KCUser CR and on that there is a UID
+// representing the user id on the user CR
+func (r *Reconciler) findUser(ctx context.Context, serverClient k8sclient.Client, loginEventUserId kccommon.Users, users []userHelper.MultiTenantUser) (*userHelper.MultiTenantUser, error) {
+	kucr, err := r.getKeycloakUserCR(ctx, serverClient, loginEventUserId)
+	if err != nil {
+		r.Log.Error("Error getting keylcoak user cr", err)
+		return nil, err
+	}
+	if kucr == nil {
+		return nil, nil
+	}
+	uid := kucr.Spec.User.FederatedIdentities[0].UserID
+	for _, user := range users {
+		if (user.UID) == uid {
+			return &user, nil
+		}
+	}
+	return nil, nil
+}
+
+func (r *Reconciler) getKeycloakUserCR(ctx context.Context, serverClient k8sclient.Client, User kccommon.Users) (*keycloak.KeycloakUser, error) {
+	var users keycloak.KeycloakUserList
+
+	listOptions := []k8sclient.ListOption{
+		k8sclient.MatchingLabels(map[string]string{
+			"sso": StagingRealmName,
+		}),
+		k8sclient.InNamespace(r.Config.GetNamespace()),
+	}
+	err := serverClient.List(ctx, &users, listOptions...)
+	if err != nil {
+		r.Log.Error("Error getting keylcoak user list", err)
+		return nil, err
+	}
+
+	for _, user := range users.Items {
+		if user.Spec.User.ID == User.UserID {
+			return &user, nil
+		}
+	}
+
+	r.Log.Warningf("Keycloak user not found", l.Fields{"UserId": User.UserID})
+
+	return nil, nil
+}
+
+func (r *Reconciler) getKeycloakUserCRByUID(ctx context.Context, serverClient k8sclient.Client, UserId string) (*keycloak.KeycloakUser, error) {
+	var users keycloak.KeycloakUserList
+
+	listOptions := []k8sclient.ListOption{
+		k8sclient.MatchingLabels(map[string]string{
+			"sso": StagingRealmName,
+		}),
+		k8sclient.InNamespace(r.Config.GetNamespace()),
+	}
+	err := serverClient.List(ctx, &users, listOptions...)
+	if err != nil {
+		r.Log.Error("Error getting keylcoak user list", err)
+		return nil, err
+	}
+
+	for _, user := range users.Items {
+		if user.Spec.User.FederatedIdentities[0].UserID == UserId {
+			return &user, nil
+		}
+	}
+
+	r.Log.Warningf("Keycloak user not found", l.Fields{"UserId": UserId})
+
+	return nil, nil
+}
+
+func (r *Reconciler) getUserIdLoginEvents(ctx context.Context, serverClient k8sclient.Client, kc *keycloak.Keycloak) ([]kccommon.Users, error) {
+
+	kcClient, err := r.KeycloakClientFactory.AuthenticatedClient(*kc)
+	if err != nil {
+		return nil, err
+	}
+	return kcClient.ListOfActivesUsersPerRealm(StagingRealmName, "", 1000)
```

Comment:
are we defaulting to 1000 realms even though we know keycloak won't be able to create it?

Reply:
1000 login events. That should be ok?